lib/ddtrace/transport/io/client.rb

```diff
@@ -0,0 +1,62 @@
+require 'ddtrace/transport/statistics'
+require 'ddtrace/transport/io/response'
+
+module Datadog
+  module Transport
+    module IO
+      # Encodes and writes tracer data to IO
+      class Client
+        include Transport::Statistics
+
+        attr_reader \
+          :encoder,
+          :out
+
+        def initialize(out, encoder)
+          @out = out
+          @encoder = encoder
+        end
+
+        def send_request(request)
+          # Write data to IO
+          # If block is given, allow it to handle writing
+          # Otherwise use default encoding.
+          response = block_given? ? yield(out, request) : send_default_request(out, request)
+
+          # Update statistics
+          update_stats_from_response!(response)
+
+          # Return response
+          response
+        rescue StandardError => e
+          message = "Internal error during IO transport request. Cause: #{e.message} Location: #{e.backtrace.first}"
+
+          # Log error
+          if stats.consecutive_errors > 0
+            Datadog::Logger.log.debug(message)
+          else
+            Datadog::Logger.log.error(message)
+          end
+
+          # Update statistics
+          update_stats_from_exception!(e)
+
+          InternalErrorResponse.new(e)
+        end
+
+        private
+
+        def send_default_request(out, request)
```

Comment:
does this ever get called? since we pass a block in `IO::Traces`?

Reply:
would we be better off with `send_request(req, &block)` or something instead since it seems like this code is duplicated in `IO::Traces`?

Reply:
Right now, no, it doesn't because its overridden by `Traces`, but its not a duplication because `Traces` has its own rules about how to encode data. (Although I'll see if I can make them a bit more consistent.)

The reason for providing this is just a simple matter of API usage; requiring a block for `#send_request` versus having some default behavior for `#send_request`, such that you can do `send_request(request)`.

There's no hard requirement here one way or the other, and we could get rid of this function entirely if we wanted. It just makes the `Client` a little harder to use for simple cases.

Reply:
`Client` now has some default functions that `Traces` reuses, so there's a little less duplication. What remains are overrides which I think are appropriate for granting flexibility to writing to IO (for both encoding and writing), unless we decide that flexibility is not something we want.